src-srv/routes.ts

```diff
@@ -0,0 +1,227 @@
+    return typeof value === 'object' && 'payload' in value
+  }
+}
+
+/*
+ * Map directory structure under api/ to express route paths.
+ * Direcory or file names enclosed in brackets, i.e api/planning[id]/[action].ts
+ * would become "api/planning/:id/:action".
+ */
+export async function mapRoutes(apiDir: string): Promise<RouteMap> {
+  const routes: RouteMap = {}
+
+  buildRoutes(routes, apiDir)
+
+  for (const route in routes) {
+    routes[route].handlers = await importRouteHandler(routes[route].path)
+  }
+
+  return routes
+}
+
+/**
+ * Connect all route handlers found to their respective routes
+ */
+export function connectRouteHandlers(app: Application, routes: RouteMap, context: RouteInitContext): Application {
+  for (const route in routes) {
+    const routePath = path.join('/api', route)
+
+    const { GET, POST, PUT, PATCH, DELETE, WEB_SOCKET } = routes[route].handlers
+
+    if (GET) {
+      connectRouteHandler(app, routePath, GET, context)
+    }
+
+    if (POST) {
+      connectRouteHandler(app, routePath, POST, context)
+    }
+
+    if (PUT) {
+      connectRouteHandler(app, routePath, PUT, context)
+    }
+
+    if (PATCH) {
+      connectRouteHandler(app, routePath, PATCH, context)
+    }
+
+    if (DELETE) {
+      connectRouteHandler(app, routePath, DELETE, context)
+    }
+
+    if (WEB_SOCKET) {
+      connectWebsocketHandler(app, routePath, WEB_SOCKET, context)
+    }
+  }
+
+  return app
+}
+
+/*
+ * Connect an exported route handler like GET, POST, etc to a specific express path.
+ */
+function connectRouteHandler(app: Application, routePath: string, func: RouteHandler, initContext: RouteInitContext): Application {
+  const handlerFunc = (req, res): void => {
+    const context: RouteContext = {
+      ...initContext,
+      res
+    }
+
+    func(req, context).then((response) => {
+      if (ApiResponse.isContent(response)) {
+        const { statusCode = 200, payload = '' } = response
+
+        res.status(statusCode).json(payload)
+      } else if (ApiResponse.isStatus(response)) {
+        const { statusCode, statusMessage } = response
+
+        res.status(statusCode).json({
+          statusCode,
+          statusMessage
+        })
+      } else {
+        res.status(500).json({
+          statusCode: 500,
+          statusMessage: 'Incorrect response from route handler. Expected payload or statusCode plus statusMessage.'
+        })
+      }
+    }).catch(ex => {
+      console.error(ex)
+      res.statusCode = 500
+      res.statusMessage = ex?.message || 'Uknown error'
+      res.send('')
+    })
+  }
+
+  switch (func.name) {
+    case 'GET':
+      return app.get(routePath, handlerFunc)
+
+    case 'POST':
+      return app.post(routePath, handlerFunc)
+
+    case 'PUT':
+      return app.put(routePath, handlerFunc)
+
+    case 'PATCH':
+      return app.patch(routePath, handlerFunc)
+
+    case 'DELETE':
+      return app.delete(routePath, handlerFunc)
+
+    // case 'WEB_SOCKET':
```

Comment:
Do we need dynamic ws-routes? Problems or how come isn't this used?

Reply:
Leftover comment, I'll remove it. Websockets are handled in another function below as it works differently.